### PR TITLE
Optimizing ColPali for Retrieval at Scale

### PR DESCRIPTION
**Pull request by sabrinaaquino**

### CODE REVIEW

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+### Pooling
+
+Pooling is well-known in machine learning as a way to compress data while keeping important information intact. For ColPali, we reduced ~1,030 vectors per page to just 38 vectors by pooling rows in the document's 32x32 grid.
+
+The most popular types of pooling are:
+
+- **Mean Pooling:** Averages values across rows.
+- **Max Pooling:** Selects the maximum value for each feature.
+
+32 vectors represent the pooled rows, while an additional 6 vectors encode contextual information derived from ColPali’s special tokens (e.g., <bos> for the beginning of the sequence, and task-specific instructions like “Describe the image”).
```
[the remark this answers is not in the document]
Comment (sabrinaaquino) [reply qdrant-landing/content/blog/colpali-optimization.md:62]: Last of 38 of 1030?

[the remark this answers is not in the document]
Comment (mrscoopers) [reply qdrant-landing/content/blog/colpali-optimization.md]: Both, actually
I take last 6 vectors out of 1030, do mean/max pooling on 1024, converting them to 32, and then append 6 reamining ones, making 38 in total

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 2]
```diff
@@ -30,15 +30,22 @@ Here's how we solved these challenges to make ColPali 13x faster without sacrifi
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:30]: We can add also pdf, vrag, visual RAG

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 3]
```diff
@@ -30,15 +30,22 @@ Here's how we solved these challenges to make ColPali 13x faster without sacrifi
...
-- **Vector Explosion:** Each page generates ~1,030 vectors of 128 dimensions.
-- **Index Complexity:** Trillions of comparisons to build the index comparing every vector pair.
+- **Number of Vectors:** Each page generates ~1,000 vectors of 128 dimensions.
 
-Even advanced indexing algorithms like **HNSW** struggle with this scale, as computational costs grow quadratically. 
+The total number of comparisons is calculated as:
+
+$$
+1,000 \cdot 1,000 \cdot 20,000 \cdot 128 = 2.56 \times 10^{16} \text{ comparisons!}
+$$
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:44]: 10^12

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 4]
```diff
@@ -30,15 +30,22 @@ Here's how we solved these challenges to make ColPali 13x faster without sacrifi
...
+
+
+
+That's trillions of comparisons needed to build the index. Even advanced indexing algorithms like **HNSW** struggle with this scale, as computational costs grow quadratically.
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:49]: quadratically... with amount of multivectors per page

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 5]
```diff
@@ -54,54 +61,52 @@ For those eager to explore, the [codebase is available here](https://github.com/
 
 Pooling is well-known in machine learning as a way to compress data while keeping important information intact. For ColPali, we reduced ~1,030 vectors per page to just 38 vectors by pooling rows in the document's 32x32 grid.
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:62]: I'd remove "intact"

Comment (mrscoopers) [reply qdrant-landing/content/blog/colpali-optimization.md:62]: Also it's exactly 1030 per page

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 6]
```diff
@@ -54,54 +61,52 @@ For those eager to explore, the [codebase is available here](https://github.com/
...
 
-The most popular types of pooling are:
+Max and mean pooling are the two most popular types, so we decided to test both approaches on the rows of the grid. Likewise, we could apply pooling on columns, which we plan to explore in the future.
 
 - **Mean Pooling:** Averages values across rows.
 - **Max Pooling:** Selects the maximum value for each feature.
 
-32 vectors represent the pooled rows, while an additional 6 vectors encode contextual information derived from ColPali’s special tokens (e.g., <bos> for the beginning of the sequence, and task-specific instructions like “Describe the image”).
+32 vectors represent the pooled rows, while the final 6 vectors encode contextual information derived from ColPali’s special tokens (e.g., <bos> for the beginning of the sequence, and task-specific instructions like “Describe the image”).
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:69]: I think "final" might be misunderstood as "we have 6 vectors as a final result"

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 7]
```diff
@@ -54,54 +61,52 @@ For those eager to explore, the [codebase is available here](https://github.com/
...
 We created a custom dataset with over 20,000 unique PDF pages by merging:
 
-- **ViDoRe Benchmark:** Designed for document retrieval evaluation.
+- **ViDoRe Benchmark:** Designed for PDF documents retrieval evaluation.
 - **UFO Dataset:** Visually rich documents paired with synthetic queries.
 - **DocVQA Dataset:** A large set of document-derived Q&A pairs.
 
 Each document was processed into 32x32 grids, generating both full-resolution and pooled embeddings. 
 
 ![](/blog/colpali-optimization/rows.png)
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:90]: I feel like maybe image should come earlier, where we talk about pooling first time

File: qdrant-landing/content/blog/colpali-optimization.md [hunk 8]
```diff
@@ -54,54 +61,52 @@ For those eager to explore, the [codebase is available here](https://github.com/
...
 
-These embeddings were stored in the **Qdrant vector database**, configured for speed:
+These embeddings were stored in the **Qdrant vector database**:
 
 - **Full-Resolution Embeddings:** ~1,030 vectors per page.
```
Comment (mrscoopers) [inline qdrant-landing/content/blog/colpali-optimization.md:94]: exactly 1030

**Review by mrscoopers — CHANGES_REQUESTED** [review]

File: qdrant-landing/content/blog/colpali-optimization.md (hunk 2 above)
Comment (sabrinaaquino) [reply qdrant-landing/content/blog/colpali-optimization.md:30]: We'll have a new blog for the RAG

**Review by mrscoopers — APPROVED** [review]

**Pull request merged**